AB#2685 Add documentation for automatic IAM creation (#711)
edgelesssys/constellation: docs/docs/getting-started/first-steps.md

2. Fill in your cloud provider specific information.

    First you need to create an [IAM configuration](../workflows/config.md#creating-an-iam-configuration). The easiest way to do this is the following CLI command:

    <tabs groupId="csp">
    <tabItem value="azure" label="Azure (CLI)">

>     You need several resources for the cluster. You can use the following `az` script to create them:

Reviewer: There's a tip at the end of this page regarding this az script, which doesn't make sense anymore and should be removed.

2. Fill in your cloud provider specific information.

    <tabs groupId="csp">

    <tabItem value="azure" label="Azure">

    ```bash
    constellation iam create azure --region=westus --resourceGroup=constellTest --servicePrincipal=spTest
    ```

    :::info

>     This command creates IAM configuration on the Azure region `westus` using a new resource group `constellTest` and a new service principal `spTest`.

Reviewer 2: using would imply to me that I am responsible for creating the SP so that `iam create` can use it.
```suggestion
    This command creates IAM configuration on the Azure region `westus` creating a new resource group `constellTest` and a new service principal `spTest`.
```

    :::

    </tabItem>

    <tabItem value="gcp" label="GCP">

    ```bash
    constellation iam create gcp --projectID=yourproject-12345 --zone=europe-west1-a --serviceAccountID=constell-test
    ```

    :::info

>     This command creates IAM configuration in the GCP project `yourproject-12345` on the GCP zone `europe-west1-a` using a new service account `constell-test`.

Reviewer 2: ```suggestion
    This command creates IAM configuration in the GCP project `yourproject-12345` on the GCP zone `europe-west1-a` creating a new service account `constell-test`.
```

2. Fill in your cloud provider specific information.

    <tabs groupId="csp">

    <tabItem value="azure" label="Azure">

    ```bash
    constellation iam create azure --region=westus --resourceGroup=constellTest --servicePrincipal=spTest
    ```

    :::info
    This command creates IAM configuration on the Azure region `westus` creating a new resource group `constellTest` and a new service principal `spTest`.
    :::

    :::tip

>     Since `clientSecretValue` is a sensitive value, you can leave it empty in the configuration file and pass it via an environment variable instead. To this end, create the environment variable `CONSTELL_AZURE_CLIENT_SECRET_VALUE` and set it to the secret value.

Author: You mean to duplicate "Fill in your cloud provider specific information" into all of the tabs?

>     Since `clientSecretValue` is a sensitive value, you can leave it empty in the configuration file and pass it via an environment variable instead. To this end, create the environment variable `CONSTELL_AZURE_CLIENT_SECRET_VALUE` and set it to the secret value.

Reviewer: Yes. Or alternatively, maybe remove the tip and have it only in the workflow? Or something else?

>     Since `clientSecretValue` is a sensitive value, you can leave it empty in the configuration file and pass it via an environment variable instead. To this end, create the environment variable `CONSTELL_AZURE_CLIENT_SECRET_VALUE` and set it to the secret value.

Author: I think removing it here is the way to go. I was unsure because it was a tip with security-impact, but i assume people who take enough care to not put the bare secret into the config file will also read the workflow page

2. Fill in your cloud provider specific information.

>     <tabs groupId="csp">

Reviewer: Yes, I'd also suggest to have the three new (and recommended) ways in sync with the tabs in step 1. So rename the value `azure-command` to `azure` and so on. You then need to rename the existing `aws` to something else.

    <tabItem value="azure-command" label="Azure">

    Create the IAM configuration on Azure with the following command (You must be authenticated with the [Azure CLI](https://learn.microsoft.com/en-us/cli/azure/install-azure-cli) in the shell session):

    ```bash

>     constellation iam create azure --region=westus --resourceGroup=constellTest --servicePrincipal=spTest

Reviewer 2: Even though the reference contains detailed information and is linked, I think we could state briefly what is expected of the user.

When reading the command for the first time, I asked myself the following question: Is `resourceGroup` a name for something that is to be generated, or a reference to something that already exists. The same for `servicePrincipal`.

Reviewer: I wonder whether we should have a workflow page for IAM creation and link to this instead. But it might also be too redundant and superfluous. Opinions?

Author: I would prefer to only briefly state the command here for users who want to set up really quickly here and moving detailed descriptions to a workflow page. There, we could also explain how the IAM resources can be re-used or deleted. Let me know your opinions.

Reviewer: > moving detailed descriptions to a workflow page. There, we could also explain how the IAM resources can be re-used or deleted

A workflow page seems a good solution if we actually have additional infos there like you propose 👍

Author: Should we also link to the workflow pages in the getting-started guide? Currently, i think we only do this for the verification section (which is removed in v2.3 anyways?) and cluster recovery. However, i think for a user it might be nice to get a link to detailed information about the step he is currently performing, such as IAM or cluster creation. What do you think?

Reviewer 2: If we have a dedicated workflow page for IAM it is probably better to link that than the CLI reference.

I think links to workflows are helpful if they do not interrupt the flow of the guide. I checked our quick start for this:

I like: "This will allow you to [recover your cluster](https://constellation-docs.edgeless.systems/constellation/workflows/recovery) in case of a disaster."

Less so: "For details, see the [verification section](https://constellation-docs.edgeless.systems/constellation/workflows/verify-cluster)."

Reviewer: completely agree with Fabian

    ```

>     Find an exact command description in our [CLI reference](https://docs.edgeless.systems/constellation/reference/cli).

Reviewer: ```suggestion
    Find an exact command description in the [CLI reference](../reference/cli.md).
```

    appClientID: $(jq -r '.appId' azureServiceAccountKey.json)
    clientSecretValue: $(jq -r '.password' azureServiceAccountKey.json)"
    constellation iam create azure --region=westus --resourceGroup=constellTest --servicePrincipal=spTest
    ```

    Fill the values produced by the script into your configuration file.
    Find an exact command description in our [CLI reference](https://docs.edgeless.systems/constellation/reference/cli).

    :::tip
    By default, Constellation uses `Standard_DC4as_v5` CVMs (4 vCPUs, 16 GB RAM) to create your cluster. Optionally, you can switch to a different VM type by modifying **instanceType** in the configuration file. For CVMs, any VM type with a minimum of 4 vCPUs from the [DCasv5 & DCadsv5](https://docs.microsoft.com/en-us/azure/virtual-machines/dcasv5-dcadsv5-series) or [ECasv5 & ECadsv5](https://docs.microsoft.com/en-us/azure/virtual-machines/ecasv5-ecadsv5-series) families is supported.

    Alternatively, you can leave `clientSecretValue` empty and provide the secret via the `CONSTELL_AZURE_CLIENT_SECRET_VALUE` environment variable.
    You can also run `constellation config instance-types` to get the list of all supported options.

>     :::

Author: The IAM create command is not going to set this environment variable, should I rework the command to do that?

>     :::

Author: I think it sounds like something which would be set automatically, if we say "**the** `CONSTELL_AZURE_CLIENT_SECRET_VALUE`". 
Also i think we should explain why an environment variable is preferred here.

I would phrase it like "Since this is a sensitive value, alternatively you can leave `clientSecretValue` empty and provide the secret via an environment variable such as `CONSTELL_AZURE_CLIENT_SECRET_VALUE`."

Probably you can improve that, but generally I would provide this information to the user, since i personally didn't get the grasp of it at first when reading it myself.

>     :::

Reviewer: Stating the reason makes sense!

"an environment variable such as" would be wrong because it's about exactly this variable:
https://github.com/edgelesssys/constellation/blob/be01cf7129d6f5ebcc87244aec8f3a94dc6c2d89/internal/constants/constants.go#L133-L135

Feel free to rephrase if you can make this more clear.

>     :::

Author: Got it. Makes sense.

"Since this is a sensitive value, alternatively you can set the `CONSTELL_AZURE_CLIENT_SECRET_VALUE` environmental variable, leave `clientSecretValue` empty in the `constellation-conf.yml` and provide the secret via the created environment variable."

Does that sound good?

>     :::

Reviewer: IMO still sounds a bit confusing. Maybe this is clearer (though a bit verbose)?

"Since this is a sensitive value, alternatively you can leave `clientSecretValue` empty in the configuration file and pass it via an environment variable instead. To this end, create the environment variable `CONSTELL_AZURE_CLIENT_SECRET_VALUE` and set it to the secret value."

>     :::

Author: Sounds good!

>     </tabItem>

Reviewer: I think this sentence belongs just below the bash command

Author: I changed that. I also noticed we didn't mention `constellation config instance-types` in every CP's tab. Should we do so? On that, I also noticed we don't mention supported AWS machines anywhere. I think we should add them somewhere, right?

Reviewer: Yes, we should do both. AWS types are probably too much to list them all, but we should at least link to https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/enable-nitrotpm-prerequisites.html

Author: Added that [here](https://github.com/edgelesssys/constellation/pull/711/commits/d71ccef4ffd52869affcaa61a24af3a80ee9d0f0). Could you verify it for technical correctness?

Reviewer: That's fine 👍

    ```

    Find an exact command description in our [CLI reference](https://docs.edgeless.systems/constellation/reference/cli).

>     Please note that currently, Constellation OS images are currently replicated to the following regions:

Reviewer: ```suggestion
    Constellation OS images are currently replicated to the following regions:
```

      Supported are all machines from the N2D family. It defaults to `n2d-standard-4` (4 vCPUs, 16 GB RAM), but you can use any other VMs from the same family. Refer to [N2D machine series](https://cloud.google.com/compute/docs/general-purpose-machines#n2d_machines) or run `constellation config instance-types` to get the list of all supported options.

    </tabItem>

>     <tabItem value="aws" label="AWS (Console)">

Reviewer: This doesn't use the AWS Console. We should name it differently, maybe "(manually)".
